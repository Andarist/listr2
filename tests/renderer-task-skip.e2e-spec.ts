/* eslint-disable @typescript-eslint/no-empty-function */
import { mockProcessExit, mockProcessStderr, mockProcessStdout } from 'jest-mock-process'

import { Listr } from '@root'

describe('show task skip', () => {
  let mockExit: jest.SpyInstance<never, [number?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStdout: jest.SpyInstance<boolean, [string, string?, Function?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStderr: jest.SpyInstance<boolean, [string, string?, Function?]>

  process.stdout.isTTY = true

  beforeEach(async () => {
    mockExit = mockProcessExit()
    mockStdout = mockProcessStdout()
    mockStderr = mockProcessStderr()
  })

  afterEach(async () => {
    mockExit.mockRestore()
    mockStdout.mockRestore()
    mockStderr.mockRestore()
    jest.clearAllMocks()
  })

  // ws7S3nDQgIm3rqk7S8Z1z9NgWUWyqx6F
  it('should skip from internal function', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (_, task): void => {
            task.skip('I am skipping this tasks for reasons.')
          }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expect(mockStdout.mock.calls).toMatchSnapshot('ws7S3nDQgIm3rqk7S8Z1z9NgWUWyqx6F-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('ws7S3nDQgIm3rqk7S8Z1z9NgWUWyqx6F-err')
    expect(mockExit.mock.calls).toMatchSnapshot('ws7S3nDQgIm3rqk7S8Z1z9NgWUWyqx6F-exit')
  })

  // 8KLp76vGVlGdzoy4HztCYcYe2coxpO7e
  it('should skip from context', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (ctx): void => {
            ctx.skip = true
          }
        },

        {
          title: 'This task will never execute.',
          skip: (ctx): boolean => ctx.skip,
          task: (): void => {}
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expect(mockStdout.mock.calls).toMatchSnapshot('8KLp76vGVlGdzoy4HztCYcYe2coxpO7e-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('8KLp76vGVlGdzoy4HztCYcYe2coxpO7e-err')
    expect(mockExit.mock.calls).toMatchSnapshot('8KLp76vGVlGdzoy4HztCYcYe2coxpO7e-exit')
  })

  // 7IvF8C3RevPE0cdsG7QZonUN1JS26n0N
  it('should not collapse skip message', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (ctx): void => {
            ctx.skip = true
          }
        },

        {
          title: 'This task will never execute.',
          skip: (ctx): boolean => ctx.skip,
          task: (): void => {}
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true, collapseSkips: false }
      }
    ).run()

    expect(mockStdout.mock.calls).toMatchSnapshot('7IvF8C3RevPE0cdsG7QZonUN1JS26n0N-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('7IvF8C3RevPE0cdsG7QZonUN1JS26n0N-err')
    expect(mockExit.mock.calls).toMatchSnapshot('7IvF8C3RevPE0cdsG7QZonUN1JS26n0N-exit')
  })

  // BmDpgfyyKMN40Ei5uinrsuOz1b2lEqtK
  it('skip from function', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (ctx): void => {
            ctx.skip = true
          }
        },

        {
          title: 'This task will never execute.',
          skip: (ctx): string | boolean => ctx.skip ? 'I will be skipped!' : false,
          task: (): void => {}
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expect(mockStdout.mock.calls).toMatchSnapshot('BmDpgfyyKMN40Ei5uinrsuOz1b2lEqtK-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('BmDpgfyyKMN40Ei5uinrsuOz1b2lEqtK-err')
    expect(mockExit.mock.calls).toMatchSnapshot('BmDpgfyyKMN40Ei5uinrsuOz1b2lEqtK-exit')
  })

  // omE6UjDQWFXPCa7F8rNE7ByEXsllnMAB
  it('should skip without a suffix', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (_, task): void => {
            task.skip('I am skipping this tasks for reasons.')
          }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true, suffixSkips: false }
      }
    ).run()

    expect(mockStdout.mock.calls).toMatchSnapshot('omE6UjDQWFXPCa7F8rNE7ByEXsllnMAB-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('omE6UjDQWFXPCa7F8rNE7ByEXsllnMAB-err')
    expect(mockExit.mock.calls).toMatchSnapshot('omE6UjDQWFXPCa7F8rNE7ByEXsllnMAB-exit')
  })

  // c4Q9Hk2x725caX6F6qXGMgyUQh36HQls
  it('should show the original title of the task', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (_, task): void => {
            task.skip('I am skipping this tasks for reasons.')
          }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true, showSkipMessage: false }
      }
    ).run()

    expect(mockStdout.mock.calls).toMatchSnapshot('c4Q9Hk2x725caX6F6qXGMgyUQh36HQls-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('c4Q9Hk2x725caX6F6qXGMgyUQh36HQls-err')
    expect(mockExit.mock.calls).toMatchSnapshot('c4Q9Hk2x725caX6F6qXGMgyUQh36HQls-exit')
  })

  // 3n2w67J8KnAihH5KaGDGnHe2O9xawvnx
  it('should show the original title of the task when skipped with empty message', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (_, task): void => {
            task.skip()
          }
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true, showSkipMessage: false }
      }
    ).run()

    expect(mockStdout.mock.calls).toMatchSnapshot('3n2w67J8KnAihH5KaGDGnHe2O9xawvnx-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('3n2w67J8KnAihH5KaGDGnHe2O9xawvnx-err')
    expect(mockExit.mock.calls).toMatchSnapshot('3n2w67J8KnAihH5KaGDGnHe2O9xawvnx-exit')
  })
})
