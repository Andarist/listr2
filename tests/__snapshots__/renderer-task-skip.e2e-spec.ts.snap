// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show task skip should not collapse skip message: 7IvF8C3RevPE0cdsG7QZonUN1JS26n0N-err 1`] = `[]`;

exports[`show task skip should not collapse skip message: 7IvF8C3RevPE0cdsG7QZonUN1JS26n0N-exit 1`] = `[]`;

exports[`show task skip should not collapse skip message: 7IvF8C3RevPE0cdsG7QZonUN1JS26n0N-out 1`] = `
[
  [
    "‚óº This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚óº This task will execute.
‚óº This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
‚óº This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚óº This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ö† This task will never execute.
  ‚Üì This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[1A[2K[G",
  ],
  [
    "‚úî This task will execute.
‚ö† This task will never execute.
  ‚Üì This task will never execute.
",
  ],
]
`;

exports[`show task skip should show the original title of the task when skipped with empty message: 3n2w67J8KnAihH5KaGDGnHe2O9xawvnx-err 1`] = `[]`;

exports[`show task skip should show the original title of the task when skipped with empty message: 3n2w67J8KnAihH5KaGDGnHe2O9xawvnx-exit 1`] = `[]`;

exports[`show task skip should show the original title of the task when skipped with empty message: 3n2w67J8KnAihH5KaGDGnHe2O9xawvnx-out 1`] = `
[
  [
    "‚óº This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚ùØ This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚Üì This task will execute. [SKIPPED]
",
  ],
  [
    "[2K[1A[2K[G",
  ],
  [
    "‚Üì This task will execute. [SKIPPED]
",
  ],
]
`;

exports[`show task skip should show the original title of the task: c4Q9Hk2x725caX6F6qXGMgyUQh36HQls-err 1`] = `[]`;

exports[`show task skip should show the original title of the task: c4Q9Hk2x725caX6F6qXGMgyUQh36HQls-exit 1`] = `[]`;

exports[`show task skip should show the original title of the task: c4Q9Hk2x725caX6F6qXGMgyUQh36HQls-out 1`] = `
[
  [
    "‚óº This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚ùØ This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚Üì This task will execute. [SKIPPED]
",
  ],
  [
    "[2K[1A[2K[G",
  ],
  [
    "‚Üì This task will execute. [SKIPPED]
",
  ],
]
`;

exports[`show task skip should skip from context: 8KLp76vGVlGdzoy4HztCYcYe2coxpO7e-err 1`] = `[]`;

exports[`show task skip should skip from context: 8KLp76vGVlGdzoy4HztCYcYe2coxpO7e-exit 1`] = `[]`;

exports[`show task skip should skip from context: 8KLp76vGVlGdzoy4HztCYcYe2coxpO7e-out 1`] = `
[
  [
    "‚óº This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚óº This task will execute.
‚óº This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
‚óº This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚óº This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚Üì This task will never execute. [SKIPPED]
",
  ],
  [
    "[2K[1A[2K[1A[2K[G",
  ],
  [
    "‚úî This task will execute.
‚Üì This task will never execute. [SKIPPED]
",
  ],
]
`;

exports[`show task skip should skip from internal function: ws7S3nDQgIm3rqk7S8Z1z9NgWUWyqx6F-err 1`] = `[]`;

exports[`show task skip should skip from internal function: ws7S3nDQgIm3rqk7S8Z1z9NgWUWyqx6F-exit 1`] = `[]`;

exports[`show task skip should skip from internal function: ws7S3nDQgIm3rqk7S8Z1z9NgWUWyqx6F-out 1`] = `
[
  [
    "‚óº This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚ùØ This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚Üì This task will execute. [SKIPPED]
",
  ],
  [
    "[2K[1A[2K[G‚Üì I am skipping this tasks for reasons. [SKIPPED]
",
  ],
  [
    "[2K[1A[2K[G",
  ],
  [
    "‚Üì I am skipping this tasks for reasons. [SKIPPED]
",
  ],
]
`;

exports[`show task skip should skip without a suffix: omE6UjDQWFXPCa7F8rNE7ByEXsllnMAB-err 1`] = `[]`;

exports[`show task skip should skip without a suffix: omE6UjDQWFXPCa7F8rNE7ByEXsllnMAB-exit 1`] = `[]`;

exports[`show task skip should skip without a suffix: omE6UjDQWFXPCa7F8rNE7ByEXsllnMAB-out 1`] = `
[
  [
    "‚óº This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚ùØ This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚Üì This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚Üì I am skipping this tasks for reasons.
",
  ],
  [
    "[2K[1A[2K[G",
  ],
  [
    "‚Üì I am skipping this tasks for reasons.
",
  ],
]
`;

exports[`show task skip skip from function: BmDpgfyyKMN40Ei5uinrsuOz1b2lEqtK-err 1`] = `[]`;

exports[`show task skip skip from function: BmDpgfyyKMN40Ei5uinrsuOz1b2lEqtK-exit 1`] = `[]`;

exports[`show task skip skip from function: BmDpgfyyKMN40Ei5uinrsuOz1b2lEqtK-out 1`] = `
[
  [
    "‚óº This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚óº This task will execute.
‚óº This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
‚óº This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚óº This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚Üì I will be skipped! [SKIPPED]
",
  ],
  [
    "[2K[1A[2K[1A[2K[G",
  ],
  [
    "‚úî This task will execute.
‚Üì I will be skipped! [SKIPPED]
",
  ],
]
`;